packages/core/contracts/optimistic-oracle-v2/implementation/WhitelistOptimisticOracleV2.sol

```diff
@@ -0,0 +1,413 @@
+            whitelist = defaultProposerWhitelist;
+        }
+
+        require(whitelist.isOnWhitelist(proposer) || msg.sender == owner(), "Proposer not whitelisted");
+        return super.proposePriceFor(proposer, requester, identifier, timestamp, ancillaryData, proposedPrice);
+    }
+
+    /**
+     * @notice Gets the custom proposer whitelist for a request.
+     * @dev This omits the timestamp from the key derivation, so the whitelist might have been set in advance of the
+     * request.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @return AddressWhitelistInterface the custom proposer whitelist for the request or zero address if not set.
+     */
+    function getCustomProposerWhitelist(
+        address requester,
+        bytes32 identifier,
+        bytes memory ancillaryData
+    ) external view returns (AddressWhitelistInterface) {
+        return customProposerWhitelists[_getId(requester, identifier, 0, ancillaryData)];
+    }
+
+    /**
+     * @notice Gets the internal request ID for a price request (without timestamp).
+     * @dev This is just a helper function that offchain systems can use for tracking the indexed
+     * CustomProposerWhitelistSet events.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @return bytes32 the request ID for the advance request.
+     */
+    function getInternalRequestId(
+        address requester,
+        bytes32 identifier,
+        bytes memory ancillaryData
+    ) external pure returns (bytes32) {
+        return _getId(requester, identifier, 0, ancillaryData);
+    }
+
+    /**
+     * @notice Sets the maximum bond that can be set for a request.
+     * @dev This can be used to limit the bond amount that can be set by request managers.
+     * @param _maximumBond new maximum bond amount.
+     */
+    function _setMaximumBond(uint256 _maximumBond) internal {
+        maximumBond = _maximumBond;
+        emit MaximumBondUpdated(_maximumBond);
+    }
+
+    /**
+     * @notice Sets the minimum liveness that can be set for a request.
+     * @dev This can be used to limit the liveness period that can be set by request managers.
+     * @param _minimumLiveness new minimum liveness period.
+     */
+    function _setMinimumLiveness(uint256 _minimumLiveness) internal {
+        minimumLiveness = _minimumLiveness;
+        emit MinimumLivenessUpdated(_minimumLiveness);
+    }
+
+    /**
+     * @notice Sets the default proposer whitelist.
+     * @param whitelist address of the whitelist to set.
+     */
+    function _setDefaultProposerWhitelist(address whitelist) internal {
+        defaultProposerWhitelist = AddressWhitelistInterface(whitelist);
```

Comment:
Should we require whitelist != zeroAddress?

Reply:
good call, we should definitely check this